src/v/cloud_topics/level_one/common/abstract_io.h

```diff
@@ -0,0 +1,98 @@
+/*
+ * Copyright 2025 Redpanda Data, Inc.
+ *
+ * Licensed as a Redpanda Enterprise file under the Redpanda Community
+ * License (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ * https://github.com/redpanda-data/redpanda/blob/master/licenses/rcl.md
+ */
+
+#pragma once
+
+#include "cloud_topics/level_one/common/object_id.h"
+#include "container/chunked_vector.h"
+
+#include <seastar/core/file.hh>
+#include <seastar/core/fstream.hh>
+#include <seastar/core/future.hh>
+#include <seastar/core/reactor.hh>
+
+#include <expected>
+#include <limits>
+
+namespace experimental::cloud_topics::l1 {
+
+// An abstraction for a local file that is used for staging uploads to object
+// storage.
+class staging_file {
+public:
+    staging_file() = default;
+    staging_file(const staging_file&) = delete;
+    staging_file(staging_file&&) = delete;
+    staging_file& operator=(const staging_file&) = delete;
+    staging_file& operator=(staging_file&&) = delete;
+    virtual ~staging_file() = default;
+
+    // Return the size of this file in bytes on disk.
+    virtual ss::future<size_t> size() = 0;
+    // Return an output stream for appending to this file.
+    virtual ss::future<ss::output_stream<char>> output_stream() = 0;
+    // Remove the file from the local filesystem.
+    //
+    // This should always be called (outside of unclean shutdown).
+    virtual ss::future<> remove() = 0;
+
+private:
+    friend class io;
+    // Return an input stream for reading from this file.
+    virtual ss::future<ss::input_stream<char>> input_stream() = 0;
+};
+
+// An abstraction for IO in level one.
+class io {
+public:
+    enum class errc : uint8_t {
+        file_io_error,
+        cloud_missing_object, // when reading something does not exist
+        cloud_op_error,
+        cloud_op_timeout,
+    };
+    io() = default;
+    io(const io&) = delete;
+    io(io&&) = delete;
+    io& operator=(const io&) = delete;
+    io& operator=(io&&) = delete;
+    virtual ~io() = default;
+
+    // Create a temporary file for staging uploads into object storage.
+    //
+    // If the operation writing to the file succeeds, then the local file should
+    // be uploaded using `io::upload_file`, then deleted using
+    // `local_file::remove()`. If there is an error than `local_file::remove()`
+    // should still be called to clean up the temporary file.
+    virtual ss::future<std::expected<std::unique_ptr<staging_file>, errc>>
+    create_tmp_file() = 0;
+
+    // Upload a local file to object storage, returning the object ID that was
+    // used to identify the object in the bucket.
+    virtual ss::future<std::expected<object_id, errc>>
+    put_object(staging_file*, ss::abort_source*) = 0;
```

Comment:
In my mind, something else above the IO layer would assign an object IO to a given object, and then pass it to the IO layer. Is your mental model different from that?

I can see an argument that the caller shouldn't need to know about the object ID before sending it. OTOH, there's also an argument to be made that the object extents are gonna be easier to create as we're constructing the staging file, so having the object ID up front is more convenient

Reply:
yeah I am fine to pass in the object ID instead. Will push an update

Reply:
Done in https://github.com/redpanda-data/redpanda/compare/ea0e15dc7a52a92047c8a9a00461b68fc1b765e0..323f4e05e7328e0314848b9b2076e6a1e03562f9